dask/highlevelgraph.py

```diff
@@ -0,0 +1,211 @@
+import toolz
+
+from .utils import ignoring
+from .base import is_dask_collection
+from .compatibility import Mapping
+
+
+class HighLevelGraph(Mapping):
+    """ Task graph composed of layers of dependent subgraphs
+
+    This object encodes a Dask task graph that is composed of layers of
+    dependent subgraphs, such as commonly occurs when building task graphs
+    using high level collections like Dask array, bag, or dataframe.
+
+    Typically each high level array, bag, or dataframe operation takes the task
+    graphs of the input collections, merges them, and then adds one or more new
+    layers of tasks for the new operation.  These layers typically have at
+    least as many tasks as there are partitions or chunks in the collection.
+    The HighLevelGraph object stores the subgraphs for each operation
+    separately in sub-graphs, and also stores the dependency structure between
+    them.
+
+    Parameters
+    ----------
+    layers : Dict[str, Mapping]
+        The subgraph layers, keyed by a unique name
+    dependencies : Dict[str, Set[str]]
+        The set of layers on which each layer depends
+
+    Examples
+    --------
+
+    Here is an idealized example that shows the internal state of a
+    HighLevelGraph
+
+    >>> import dask.dataframe as dd
+
+    >>> df = dd.read_csv('myfile.*.csv')  # doctest: +SKIP
+    >>> df = df + 100  # doctest: +SKIP
+    >>> df = df[df.name == 'Alice']  # doctest: +SKIP
+
+    >>> graph = df.__dask_graph__()  # doctest: +SKIP
+    >>> graph.layers  # doctest: +SKIP
+    {
+     'read-csv': {('read-csv', 0): (pandas.read_csv, 'myfile.0.csv'),
+                  ('read-csv', 1): (pandas.read_csv, 'myfile.1.csv'),
+                  ('read-csv', 2): (pandas.read_csv, 'myfile.2.csv'),
+                  ('read-csv', 3): (pandas.read_csv, 'myfile.3.csv')},
+     'add': {('add', 0): (operator.add, 'myfile.0.csv', 100),
+             ('add', 1): (operator.add, 'myfile.1.csv', 100),
+             ('add', 2): (operator.add, 'myfile.2.csv', 100),
+             ('add', 3): (operator.add, 'myfile.3.csv', 100)}
```

Comment:
Sorry trying to follow this here, should `'myfile.*.csv'` be `('read-csv', *)` or am I missing something?